java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java

```diff
@@ -559,29 +588,131 @@ void stopHardReferencingMavenPoject() {
         return newproject;
     }
 
+    /**
+     * Task that potential project reloads should wait on. If set, a {@link fireProjectReload}(true) will be scheduled only after this blocker finishes.
+     */
+    private RequestProcessor.Task reloadBlocker;
 
-
+    /**
+     * Schedules project operation that delays potential reloads. If a reload is posted, it will be performed only after
+     * this operation compeltes (successfully, or erroneously). Multiple project operations can be scheduled, an eventual project reload
+     * should happen after all those operations complete. It is possible to postpone project reload indefinitely, avoid unnecessary
+     * operation schedules.
+     * <p>
+     * To avoid race condition on task startup, this method actually creates and schedules the task so it blocks reloads from its inception.
+     * It returns the value of the worker task as the result value. 
+     * wrapper.
+     * @param rp request processor that should schedule the task
+     * @param delay optional delay, use 0 for immediate run
+     * @param r operation to run
+     * @return the scheduled task
+     */
+    public RequestProcessor.Task scheduleProjectOperation(RequestProcessor rp, Runnable r,  int delay) {
+        RequestProcessor.Task t = rp.create(r);
+        if (Boolean.getBoolean("test.reload.sync")) {
+            LOG.log(Level.FINE, "Running the blocking task synchronously (test.reload.sync set)");
+            t.run();
+            return t;
+        } else {
+            synchronized (this) {
+                if (reloadBlocker == null) {
+                    LOG.log(Level.FINER, "Blocking project reload on task {0}", t);
+                    reloadBlocker = t;
+                } else {
+                    // will chain after existing reload blocker AND the new task.
+                    final RequestProcessor.Task t2 = RELOAD_RP.create(() -> {});
+                    LOG.log(Level.FINER, "Creating project blocker {0}, chain after existing blocker {1}", new Object[] { t2, t });
+                    reloadBlocker.addTaskListener((e) -> {
+                        t.addTaskListener((e2) -> {
+                            synchronized (NbMavenProjectImpl.this) {
+                                if (t2 == reloadBlocker) {
+                                    reloadBlocker = null;
+                                }
+                            }
+                            t2.run();
+                        });
+                    });
+                    reloadBlocker = t2;
+                }
+            }
+            t.schedule(delay);
+            return t;
```

Comment:
Ok the events are fired even if the listener is added _after_ a task finishes - thats good. I probably would leave the delay parameter out, since this blocks the reload queue while doing nothing if I understand this correctly.

Are project ops allowed to run in parallel? If yes then `reloadBlocker` won't be sufficient as marker, since there can be multiple blockers at the same time and they can finish in arbitrary order.

I wonder if this would be simpler with a queue instead of events.

Reply:
> Are project ops allowed to run in parallel? If yes then `reloadBlocker` won't be sufficient as marker, since there can be multiple blockers at the same time and they can finish in arbitrary order.
> 

Who knows, I've just invented that concept ... but I'd say yes: there may be many operations having known (destructive :)) impact on loaded metadata, so it makes sense to delay the reloads.

Re. sufficiency: Pls. elaborate, I am not sure I've got the point.

> I wonder if this would be simpler with a queue instead of events. Since this listener chain essentially imitates a queue.

Pls. elaborate -- do you think something like:
- have a List<Task>, each having a TaskListener
- listener removes its Task from list, checks emptiness and if so, schedules a reload ?

Reply:
Lets say one thread runs the ops sequentially, then tasks could be simply put into a single queue: `op0, op1, reload, op2`

This could even optimize the queue by only allowing one reload. If a second one is posted the other reload op is removed, so that it is always at the end of the queue.

If ops run in parallel however:
```
|--op0------------|
|--op1----|
|--op2-------|
```
and someone schedules a reload, which task is the blocker? This is difficult to track with an event system and one single variable.

Reply:
i can take a closer look at it this weekend. but you could also queue all tasks and also the reload task on one sequential RP with 1 thread. It is probably the safest and simplest approach. If this turns out to be a performance bottleneck we can think about parallelizing this later.

Reply:
added a commit on top of this PR: https://github.com/mbien/netbeans/commit/7d40cf67b0b2dd4077096d02bc4d067a4e57f5e7

While doing that i noticed another thing what is likely a bug: the reload task is set to `finished` during construction, but there are places in the code which check if its finished, for example `getFreshOriginalMavenProject` which would essentially always return true and the else branch would never run.

(I don't know all the details of NB's concurrency classes, so this might just be me not understanding something, however I changed that in the commit too since i couldn't find the spot which would set the task to not finished when it starts running)